src/main/java/com/google/sps/servlets/PhotoHandlerServlet.java

```diff
@@ -0,0 +1,152 @@
+
+import org.jsoup.nodes.Document;
+
+/**
+ * When the user submits the form for image uploading, Blobstore processes the file upload and 
+ * forwards the request to this servlet, which gets the shopping results for the respective photo.
+ */
+@WebServlet("/handle-photo")
+public class PhotoHandlerServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the BlobKey that points to the image uploaded by the user.
+    BlobKey blobKey = getBlobKey(request, "photo");
+
+    // Send an error if the user did not upload a file.
+    if (blobKey == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Client must upload an image file.");
+    }
+
+    // Get the photo category (i.e. product, shopping-list or barcode) entered by the user.
+    if (request.getParameter("photo-category").isEmpty()) {
+      response.sendError(
+          HttpServletResponse.SC_BAD_REQUEST, "Client must select a photo category when submitting the form.");
+    }
+    String photoCategory = request.getParameter("photo-category");
+
+    // Get the image the user uploaded as bytes.
+    byte[] imageBytes = getBlobBytes(blobKey);
+
+    // The shopping results page is returned as the response, therefore set its content type
+    // to "text/html".
+    response.setContentType("text/html");
+
+    // Based on the {@code photoCategory}, call methods from the photo detection classes, 
+    // passing {@code imageBytes} as argument. These methods detect the image content, build 
+    // the shopping query and call GoogleShoppingQuerier to return shopping results.
+
+    if (photoCategory.equals("product")) {
+      // Initialize the ProductPhotoShoppingImpl object.
+      ProductPhotoShoppingImpl productPhotoShoppingImpl = new ProductPhotoShoppingImpl();
+
+      try {
+        // Get the shopping results page as jsoup Document.
+        Document shoppingResultsDocument = productPhotoShoppingImpl.shopWithPhoto(imageBytes);
+        response.getWriter().println(shoppingResultsDocument);
+      } catch(IllegalArgumentException | 
+              ShoppingQuerierConnectionException | 
+              ProductPhotoShoppingException | 
+              IOException exception) {
+        response.sendError(SC_INTERNAL_SERVER_ERROR, exception.getMessage());
+      }
+    }
+  }
+
+  /**
+   * Returns the BlobKey corresponding to the file uploaded by the user, or null if the user did not
+   * upload a file.
+   */
+  private BlobKey getBlobKey(HttpServletRequest request, String formFileInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formFileInputElementName);
+
+    // User submitted form without selecting a file. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
```

Comment:
Prefer to use "Optional.empty()" instead of null.

Reply:
What a nice suggestion, thanks, done!